File: Sources/Encryption.swift
import Foundation

/// SRUN3K 加密模块
/// 移植自 ehaut 项目的 JavaScript 加密算法
struct SrunEncryption {

    /// 加密用户名
    /// 算法：每个字符 ASCII +4，前缀 "{SRUN3}\r\n"
    static func encryptUsername(_ username: String) -> String {
        var encrypted = ""
        for char in username {
            if let scalar = char.unicodeScalars.first {
                let newValue = scalar.value + 4
                if let newScalar = UnicodeScalar(newValue) {
                    encrypted += String(newScalar)
                }
            }
        }
        return "{SRUN3}\r\n" + encrypted
    }

    /// 加密密码
    /// 算法：XOR 加密，密钥 "1234567890"，按位拆分后交替组合
    static func encryptPassword(_ password: String) -> String {
        let key = "1234567890"
        let keyChars = Array(key)
        var encrypted = ""

        for (i, char) in password.enumerated() {
            guard let charValue = char.asciiValue else { continue }

            // 计算密钥索引：key.length - 1 - (i % key.length)
            let keyIndex = key.count - 1 - (i % key.count)
            guard let keyCharValue = keyChars[keyIndex].asciiValue else { continue }

            // XOR 运算
            let ki = Int(charValue) ^ Int(keyCharValue)

            // 拆分为低4位和高4位
            let lowBits = (ki & 0x0f) + 0x36
            let highBits = ((ki >> 4) & 0x0f) + 0x63

            let lowChar = Character(UnicodeScalar(lowBits)!)
            let highChar = Character(UnicodeScalar(highBits)!)

            // 根据索引奇偶交替组合
            if i % 2 == 0 {
                encrypted += String(lowChar) + String(highChar)
            } else {
                encrypted += String(highChar) + String(lowChar)
            }
        }

        return encrypted
    }
}

File: Windows-Rust-Deprecated/src/encryption.rs
//! SRUN3K 加密算法模块

/// 加密用户名: ASCII + 4, 前缀 {SRUN3}\r\n
pub fn encrypt_username(username: &str) -> String {
    let encrypted: String = username
        .chars()
        .map(|c| char::from_u32(c as u32 + 4).unwrap_or(c))
        .collect();
    format!("{{SRUN3}}\r\n{}", encrypted)
}

/// 加密密码: XOR + 位拆分
/// 密钥: "1234567890", 反向索引
/// 低4位 + 0x36, 高4位 + 0x63
/// 偶数索引: low+high, 奇数索引: high+low
pub fn encrypt_password(password: &str) -> String {
    let key = b"1234567890";
    let key_len = key.len();
    let mut encrypted = String::new();

    for (i, c) in password.bytes().enumerate() {
        let key_index = key_len - 1 - (i % key_len);
        let ki = c ^ key[key_index];

        let low_bits = (ki & 0x0f) + 0x36;
        let high_bits = ((ki >> 4) & 0x0f) + 0x63;

        let low_char = char::from(low_bits);
        let high_char = char::from(high_bits);

        if i % 2 == 0 {
            encrypted.push(low_char);
            encrypted.push(high_char);
        } else {
            encrypted.push(high_char);
            encrypted.push(low_char);
        }
    }

    encrypted
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_encrypt_username() {
        let result = encrypt_username("231040600203");
        assert!(result.starts_with("{SRUN3}\r\n"));
    }

    #[test]
    fn test_encrypt_password() {
        let result = encrypt_password("test123");
        assert!(!result.is_empty());
    }
}
